### DISCUSSION TITLE
Why does xmake.lua in parent directory change project behavior in current directory?

### DISCUSSION THREAD

**Discussion by bigmodem** [xmake-io/xmake#4418] [Q&A]
### Xmake 版本

2.8.5

### 操作系统版本和架构

Mac OSX Sonoma 14.1.1 

### 描述问题

```console
%mkdir a
%cd a
%xmake create -l c -P ./b

create b ...
  [+]: xmake.lua
  [+]: src/main.c
  [+]: .gitignore
create ok!

%cd b
%xmake

checking for platform ... macosx
checking for architecture ... x86_64
checking for Xcode directory ... /Applications/Xcode.app
checking for Codesign Identity of Xcode ... 
checking for SDK version of Xcode for macosx (x86_64) ... 14.0
checking for Minimal target version of Xcode for macosx (x86_64) ... 14.0
[ 25%]: cache compiling.release src/main.c
[ 50%]: linking.release b
[100%]: build ok, spent 0.896s

%touch ../xmake.lua
%xmake
error: @programdir/core/base/interpreter.lua:518: the scope root() is empty!
stack traceback:
    [C]: in function 'error'
    [@programdir/core/base/os.lua:923]:
    [@programdir/core/base/interpreter.lua:518]:
    [C]: in function 'xpcall'
    [@programdir/core/base/interpreter.lua:751]: in function 'make'
    [@programdir/core/project/project.lua:237]: in function '_load_scope'
    [@programdir/core/project/project.lua:201]: in function '_load'
    [@programdir/core/project/project.lua:420]: in function '_load_options'
    [@programdir/core/project/project.lua:1107]:
    [C]: in function 'xpcall'
    [@programdir/core/base/utils.lua:280]:
    [@programdir/core/base/task.lua:91]: in function '_translate_menu'
    [@programdir/core/base/task.lua:465]: in function 'taskmenu'
    [@programdir/core/base/option.lua:358]: in function 'taskmenu'
    [@programdir/core/base/option.lua:444]: in function 'defaults'
    [@programdir/core/base/option.lua:86]: in function 'save'
    [@programdir/core/sandbox/modules/import/core/base/task.lua:48]: in function 'run'
    [@programdir/actions/build/main.lua:191]:
    [C]: in function 'xpcall'
    [@programdir/core/base/utils.lua:280]:
    [@programdir/core/base/task.lua:501]: in function 'run'
    [@programdir/core/main.lua:307]: in function 'cotask'
    [@programdir/core/base/scheduler.lua:404]:

checking for Xcode directory ... /Applications/Xcode.app
checking for Codesign Identity of Xcode ... 
checking for SDK version of Xcode for macosx (x86_64) ... 14.0
checking for Minimal target version of Xcode for macosx (x86_64) ... 14.0
error: @programdir/core/base/interpreter.lua:518: the scope root() is empty!
stack traceback:
    [C]: in function 'error'
    [@programdir/core/base/os.lua:923]:
    [@programdir/core/base/interpreter.lua:518]:
    [C]: in function 'xpcall'
    [@programdir/core/base/interpreter.lua:751]: in function 'make'
    [@programdir/core/project/project.lua:237]: in function '_load_scope'
    [@programdir/core/project/project.lua:201]: in function '_load'
    [@programdir/core/project/project.lua:420]: in function '_load_options'
    [@programdir/core/project/project.lua:950]: in function 'options'
    [...dir/core/sandbox/modules/import/core/project/project.lua:79]: in function 'check_options'
    [@programdir/actions/config/main.lua:366]:
    [C]: in function 'xpcall'
    [@programdir/core/base/utils.lua:280]:
    [@programdir/core/base/task.lua:501]: in function 'run'
    [@programdir/core/sandbox/modules/import/core/base/task.lua:63]: in function 'run'
    [@programdir/actions/build/main.lua:191]:
    [C]: in function 'xpcall'
    [@programdir/core/base/utils.lua:280]:
    [@programdir/core/base/task.lua:501]: in function 'run'
    [@programdir/core/main.lua:307]: in function 'cotask'
    [@programdir/core/base/scheduler.lua:404]:

warning: You are working in the project directory(/Users/xz/a) and you can also
force to build in current directory via run `xmake -P .`

%xmake -P .  
[100%]: build ok, spent 0.023s
(forcing use directory work)

%xmake -P . run 
[100%]: build ok, spent 0.013s

%xmake -P . run b
...
...
error: invalid argument: b
(forcing run target in current directory not work)

%rm ../xmake.lua
%xmake

checking for Xcode directory ... /Applications/Xcode.app
checking for Codesign Identity of Xcode ... 
checking for SDK version of Xcode for macosx (x86_64) ... 14.0
checking for Minimal target version of Xcode for macosx (x86_64) ... 14.0
[100%]: build ok, spent 0.606s

%xmake run
hello world!

%xmake run b
hello world!
```

### 期待的结果

I think project behavior in current folder shouldn't be affected by xmake.lua in parent folder, if this is just a function of xmake, then how can I disable it?

### 工程配置

add_rules("mode.debug", "mode.release")

target("b")
    set_kind("binary")
    add_files("src/*.c")

### 附加信息和错误日志

error: @programdir/core/base/interpreter.lua:518: the scope root() is empty!
stack traceback:
    [C]: in function 'error'
    [@programdir/core/base/os.lua:923]:
    [@programdir/core/base/interpreter.lua:518]:
    [C]: in function 'xpcall'
    [@programdir/core/base/interpreter.lua:751]: in function 'make'
    [@programdir/core/project/project.lua:237]: in function '_load_scope'
    [@programdir/core/project/project.lua:201]: in function '_load'
    [@programdir/core/project/project.lua:420]: in function '_load_options'
    [@programdir/core/project/project.lua:1107]:
    [C]: in function 'xpcall'
    [@programdir/core/base/utils.lua:280]:
    [@programdir/core/base/task.lua:91]: in function '_translate_menu'
    [@programdir/core/base/task.lua:465]: in function 'taskmenu'
    [@programdir/core/base/option.lua:358]: in function 'taskmenu'
    [@programdir/core/base/option.lua:444]: in function 'defaults'
    [@programdir/core/base/option.lua:86]: in function 'save'
    [@programdir/core/sandbox/modules/import/core/base/task.lua:48]: in function 'run'
    [@programdir/actions/build/main.lua:191]:
    [C]: in function 'xpcall'
    [@programdir/core/base/utils.lua:280]:
    [@programdir/core/base/task.lua:501]: in function 'run'
    [@programdir/core/main.lua:307]: in function 'cotask'
    [@programdir/core/base/scheduler.lua:404]:

error: @programdir/core/base/interpreter.lua:518: the scope root() is empty!
stack traceback:
    [C]: in function 'error'
    [@programdir/core/base/os.lua:923]:
    [@programdir/core/base/interpreter.lua:518]:
    [C]: in function 'xpcall'
    [@programdir/core/base/interpreter.lua:751]: in function 'make'
    [@programdir/core/project/project.lua:237]: in function '_load_scope'
    [@programdir/core/project/project.lua:201]: in function '_load'
    [@programdir/core/project/project.lua:985]: in function 'requires_str'
    [@programdir/core/project/project.lua:494]: in function '_load_requires'
    [@programdir/core/project/project.lua:969]: in function 'required_packages'
    [@programdir/core/project/project.lua:339]: in function '_load_targets'
    [@programdir/core/project/project.lua:909]: in function 'targets'
    [@programdir/actions/config/main.lua:115]: in function '_check_targets'
    [@programdir/actions/config/main.lua:394]:
    [C]: in function 'xpcall'
    [@programdir/core/base/utils.lua:280]:
    [@programdir/core/base/task.lua:501]: in function 'run'
    [@programdir/core/sandbox/modules/import/core/base/task.lua:63]: in function 'run'
    [@programdir/actions/build/main.lua:191]:
    [C]: in function 'xpcall'
    [@programdir/core/base/utils.lua:280]:
    [@programdir/core/base/task.lua:501]: in function 'run'
    [@programdir/core/main.lua:307]: in function 'cotask'
    [@programdir/core/base/scheduler.lua:404]:

**Comment by waruqi** [comment]
The top level directory where xmake.lua is located is the root directory of the project, and the user can build it by executing xmake from any subdirectory.  like git

if you want to build the seperate sub project. you can run `xmake build -P subdir` or move it to other directory.

**Reply by bigmodem** [reply]
Thanks for the quick response!

I just want to try let each sub project use different architecture.
For example:

server/src                              --- arch neutral source
server/xmake.lua                  --- build all sub projects, run any selected or all targets, tests in any or all sub projects 

server/armv7/src                   --- armv7 related source
server/armv7/xmake.lua.      --- build armv7 related targets, run any selected or all armv7 related targets, tests

I still have these sub projects:

server/aarch64
server/riscv32
server/riscv64
  
Can I put all arch related code in sub project and put all arch neutral code in parent project? How to config master project and sub projects for cross compile? I tried many times but always fail.

Does xmake support different arch for different sub project? Could you give me a sample ?

**Reply by waruqi** [reply]
you can use `includes` to contain all subproject's xmake.lua and run xmake command in root directory.

server/xmake.lua
 
```lua
includes("armv7", "aarch64", "riscv32", "riscv64")
```

then run `xmake` to build all targets.

```console
xmake
```

or build the given target

```console
xmake build armv7_target
```

**Reply by bigmodem** [reply]
Sorry, it failed again. 
Follow your advice, I put

includes("aarch64", "armv7")

in my parent xmake.lua. And it's the only content of it. 

But now matter what I do, the parent always force subject project  use default arch -- x86_64, and all subject project can't cross compile with their own archs.

Could you give me a working example? 
By the way, instead of gcc, I use llvm as my cross compiler, this is the script which I used to configure my subject project:

```
xmake f \
    -m debug \
    -a armv7 \
    -p cross \
    -o ../../build \
    --toolchain=llvm \
    --sdk=/usr/local/opt/llvm \
    --ld=/usr/local/opt/llvm/bin/ld.lld
```

**Reply by waruqi** [reply]
I'm not sure what you're trying to do, `includes` just imports all the subprojects and their targets.

If you want different targets to build different architecture by default, it has nothing to do with whether it's a subproject or not. You need to modify the toolchain for each target using the set_plat/set_arch/set_toolchains configuration.

```lua
-- x86_64
target("foo")
    add_files("src/*.cpp")

toolchain("myllvm")
    set_sdkdir("/usr/local/opt/llvm")
    set_ .... -- more

-- aarch64
target("bar")
    set_plat("cross")
    set_arch("arm64")
    set_toolchains("myllvm")
    add_files("src/*.cpp")
```

Since the built-in llvm toolchain does not currently support cross-compilation, it only supports x86_64/i386 arch, so you have to define your own toolchain.

https://github.com/xmake-io/xmake/blob/51db2a04f6def44e67f23448d910da1777db3c48/xmake/toolchains/llvm/xmake.lua#L62

**Reply by waruqi** [reply]
If you only want to compile a few targets at a time for a specific architecture, you can selectively enable targets based on is_plat.

```lua
if is_arch("x86_64")
    includes("x86_64")
elseif is_arch("arm64")
    includes("arm64")
end
```

or 

```lua
includes("x86_64", "arm64")
```

```lua
target("test")
    if not is_arch("arm64") then
        set_enabled(false)
    end
```

```lua
target("test2")
    if not is_arch("x86_64") then
        set_enabled(false)
    end
```
